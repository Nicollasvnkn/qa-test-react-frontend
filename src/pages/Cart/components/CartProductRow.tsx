import {
  MdAddCircleOutline,
  MdDelete,
  MdRemoveCircleOutline,
} from "react-icons/md";

// Components
import { Table } from "@/components/Table";

// Types
import { ProductInCart } from "@/@types/application";

// Helpers
import { formatPrice } from "@/helpers/formatPrice";

import {
  InputContainer,
  SubTotalContainer,
  ImageAndTitle,
  ActionButton,
} from "./styles";

type CartProviderProps = {
  product: ProductInCart;
  handleProductDecrement: (product: ProductInCart) => void;
  handleProductIncrement: (product: ProductInCart) => void;
  handleRemoveProduct: (productId: string) => void;
  isLoading: boolean;
};

export function CartProductRow({
  product,
  handleProductDecrement,
  handleProductIncrement,
  handleRemoveProduct,
  isLoading,
}: CartProviderProps) {
  return (
    <Table.Row key={product.id}>
      <ImageAndTitle>
        <img 
          src={product.imageUrl} 
          alt={product.title} 
          data-cy={`product-image-${product.id}`} // Add data-cy here for the image
        />
        <div>
          <strong>{product.title}</strong>
          <span data-cy={`product-price-${product.id}`}>{formatPrice(product.price)}</span> // And here for the price
        </div>
      </ImageAndTitle>
      <InputContainer>
        <ActionButton
          type="button"
          data-cy={`decrement-product-${product.id}`} // Add data-cy for decrement action
          disabled={isLoading || product.quantity <= 1}
          onClick={() => handleProductDecrement(product)}
        >
          <MdRemoveCircleOutline size={20} />
        </ActionButton>
        <input
          type="text"
          data-cy={`product-quantity-${product.id}`} // Add data-cy for product quantity input
          readOnly
          value={product.quantity}
        />
        <ActionButton
          type="button"
          data-cy={`increment-product-${product.id}`} // Add data-cy for increment action
          onClick={() => handleProductIncrement(product)}
          disabled={isLoading}
        >
          <MdAddCircleOutline size={20} />
        </ActionButton>
      </InputContainer>
      <SubTotalContainer>
        <span>SUBTOTAL</span>
        <strong data-cy={`product-subtotal-${product.id}`}>{formatPrice(product.amount)}</strong> // Add data-cy for subtotal
      </SubTotalContainer>
      <ActionButton
        type="button"
        data-cy={`remove-product-${product.id}`} // Add data-cy for remove action
        onClick={() => handleRemoveProduct(product.id)}
        disabled={isLoading}
      >
        <MdDelete size={20} />
      </ActionButton>
    </Table.Row>
  );
}
