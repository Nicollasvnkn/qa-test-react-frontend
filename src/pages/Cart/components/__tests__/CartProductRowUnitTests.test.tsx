import { render, fireEvent } from '@testing-library/react';
import { CartProductRow } from '../CartProductRow'; 
import '@testing-library/jest-dom';


interface MockProduct {
  id: string;
  orderItemId: string;
  imageUrl: string;
  title: string;
  price: number;
  quantity: number;
  amount: number;
}

const mockProduct: MockProduct = {
  id: '1',
  orderItemId: 'order-item-1',
  imageUrl: 'image-url.jpg',
  title: 'Product Title',
  price: 29.99,
  quantity: 2,
  amount: 59.98
};

const mockIncrement = jest.fn();
const mockDecrement = jest.fn();
const mockRemove = jest.fn();

const formatPrice = (price: number) => {
  return `R$ ${price.toFixed(2).replace('.', ',')}`;
};

describe('CartProductRow', () => {
  it('renders the product information', () => {
    const { getByText, getByAltText, getByDisplayValue } = render(
      <CartProductRow
        product={mockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );

    expect(getByAltText(mockProduct.title)).toBeInTheDocument();
    expect(getByText(mockProduct.title)).toBeInTheDocument();
    expect(getByText(formatPrice(mockProduct.price))).toBeInTheDocument();
    expect(getByDisplayValue(mockProduct.quantity.toString())).toBeInTheDocument();
    expect(getByText(formatPrice(mockProduct.amount))).toBeInTheDocument();
  });

  it('disables decrement button when product quantity is 1', () => {
    const mockProductWithQuantity = { ...mockProduct, quantity: 1 };
  
    const { container } = render(
      <CartProductRow
        product={mockProductWithQuantity}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );
  
    const decrementButton = container.querySelector(`[data-cy="decrement-product-${mockProductWithQuantity.id}"]`);
  
    expect(decrementButton).toBeDisabled();
  });  
  

  it('disables action buttons when isLoading is true', () => {
    const { container } = render(
      <CartProductRow
        product={mockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={true}
      />
    );
  
    const incrementButton = container.querySelector(`[data-cy="increment-product-${mockProduct.id}"]`);
    const decrementButton = container.querySelector(`[data-cy="decrement-product-${mockProduct.id}"]`);
    const removeButton = container.querySelector(`[data-cy="remove-product-${mockProduct.id}"]`);
  
    expect(incrementButton).toBeDisabled();
    expect(decrementButton).toBeDisabled();
    expect(removeButton).toBeDisabled();
  });  

  it('comparar a pagina com o snapshot', () => {
    const { asFragment } = render(
      <CartProductRow
        product={mockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('chama função externa quando o botão de remover é clicado', () => {
    const { container } = render(
      <CartProductRow
        product={mockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );
  
    const removeButton = container.querySelector(`[data-cy="remove-product-${mockProduct.id}"]`);
  
    // Check if the element exists before firing the event
    if (removeButton) {
      fireEvent.click(removeButton);
      expect(mockRemove).toHaveBeenCalledWith(mockProduct.id);
    } else {
      throw new Error('Remove button not found');
    }
  });  

  it('gerencia corretamente a sequência de interações do usuário', () => {
    const { container } = render(
      <CartProductRow
        product={mockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );
  
    const incrementButton = container.querySelector(`[data-cy="increment-product-${mockProduct.id}"]`);
    const decrementButton = container.querySelector(`[data-cy="decrement-product-${mockProduct.id}"]`);
    const removeButton = container.querySelector(`[data-cy="remove-product-${mockProduct.id}"]`);
  
    if (incrementButton && decrementButton && removeButton) {
      fireEvent.click(incrementButton);
      expect(mockIncrement).toHaveBeenCalledWith(mockProduct);
  
      fireEvent.click(decrementButton);
      expect(mockDecrement).toHaveBeenCalledWith(mockProduct);
  
      fireEvent.click(removeButton);
      expect(mockRemove).toHaveBeenCalledWith(mockProduct.id);
    } else {
      throw new Error('One or more buttons not found');
    }
  });


  it('verificar se a pagina não renderiza novamente', () => {
    const { rerender } = render(
      <CartProductRow
        product={mockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );
    const newMockProduct = { ...mockProduct };

    rerender(
      <CartProductRow
        product={newMockProduct}
        handleProductIncrement={mockIncrement}
        handleProductDecrement={mockDecrement}
        handleRemoveProduct={mockRemove}
        isLoading={false}
      />
    );
  });

});

