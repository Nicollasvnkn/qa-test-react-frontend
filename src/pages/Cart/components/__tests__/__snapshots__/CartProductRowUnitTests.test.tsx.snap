// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartProductRow comparar a pagina com o snapshot 1`] = `
<DocumentFragment>
  <li
    class="sc-hKgKIp klAmza"
  >
    <div
      class="sc-jSguLX burahQ"
    >
      <img
        alt="Product Title"
        data-cy="product-image-1"
        src="image-url.jpg"
      />
      <div>
        <strong>
          Product Title
        </strong>
        <span
          data-cy="product-price-1"
        >
          R$ 29,99
        </span>
         // And here for the price
      </div>
    </div>
    <div
      class="sc-gKsecS jBPPuH"
    >
      <button
        class="sc-iBPTVF hjnMhy"
        data-cy="decrement-product-1"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M7 11v2h10v-2H7zm5-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </button>
      <input
        data-cy="product-quantity-1"
        readonly=""
        type="text"
        value="2"
      />
      <button
        class="sc-iBPTVF hjnMhy"
        data-cy="increment-product-1"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </button>
    </div>
    <div
      class="sc-pGaPU hIeSBB"
    >
      <span>
        SUBTOTAL
      </span>
      <strong
        data-cy="product-subtotal-1"
      >
        R$ 59,98
      </strong>
       // Add data-cy for subtotal
    </div>
    <button
      class="sc-iBPTVF hjnMhy"
      data-cy="remove-product-1"
      type="button"
    >
      <svg
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
  </li>
</DocumentFragment>
`;
